### PR TITLE
Bluetooth: controller: Remove redundant PPI and CC

### PR DESCRIPTION
**Pull request by ioannisg**
This commit implements the SW autoswitch for TIFS on nRF52840
using one PPI channel and one TIMER CC register less. In addition
it performs a minor adjustment on the HW timing constants for LE
Coded PHY on nRF52840, based on empirical measurements.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

### CODE REVIEW

File: subsys/bluetooth/controller/hal/nrf5/radio/radio.c [hunk 1]
```diff
@@ -379,6 +379,11 @@ static void sw_switch(u8_t dir, u8_t phy_curr, u8_t flags_curr, u8_t phy_next,
 			HAL_SW_SWITCH_RADIO_ENABLE_PPI_REGISTER_TASK(ppi_en) =
 				HAL_SW_SWITCH_RADIO_ENABLE_PPI_TASK_TX;
 
+			/* Include PPI for S2 timing in the active group */
```
Comment (ioannisg) [inline subsys/bluetooth/controller/hal/nrf5/radio/radio.c]: The concept is: have only one PPI and only one CC for S2-switch, and add the PPI in the active group (i.e. for next Switch), if we are about to receive in LE Coded.

File: subsys/bluetooth/controller/hal/nrf5/radio/radio.c [hunk 2]
```diff
@@ -404,17 +409,14 @@ static void sw_switch(u8_t dir, u8_t phy_curr, u8_t flags_curr, u8_t phy_next,
 			    HAL_SW_SWITCH_TIMER_S8_DISABLE_PPI_ENABLE;
 		} else {
 			/* Switching to TX after RX on LE 1M/2M PHY */
-			u8_t ppi_en =
-			    HAL_SW_SWITCH_RADIO_ENABLE_S2_PPI(sw_tifs_toggle);
 			u8_t ppi_dis =
 			    HAL_SW_SWITCH_GROUP_TASK_DISABLE_PPI(
 			    sw_tifs_toggle);
 
-			/* Invalidate PPI used when RXing on LE Coded PHY. */
```
Comment (ioannisg) [inline subsys/bluetooth/controller/hal/nrf5/radio/radio.c]: Reason to remove this invalidation is that the PPI will not be included in the active group.

File: subsys/bluetooth/controller/hal/nrf5/radio/radio.c [hunk 3]
```diff
@@ -440,18 +442,10 @@ static void sw_switch(u8_t dir, u8_t phy_curr, u8_t flags_curr, u8_t phy_next,
...
 #if defined(CONFIG_SOC_NRF52840)
 		if (1) {
-			u8_t ppi_en =
-				HAL_SW_SWITCH_RADIO_ENABLE_S2_PPI(
-					sw_tifs_toggle);
 			u8_t ppi_dis =
 				HAL_SW_SWITCH_GROUP_TASK_DISABLE_PPI(
 					sw_tifs_toggle);
 
-			HAL_SW_SWITCH_RADIO_ENABLE_PPI_REGISTER_EVT(
```
Comment (ioannisg) [inline subsys/bluetooth/controller/hal/nrf5/radio/radio.c]: The reason to remove the invalidation is that we are about to TX (and later switch to RX). By design of the sw_tifs_toggle, the PPI for S2 cannot be in the active group (i.e. it belongs to the "other" group)

File: subsys/bluetooth/controller/hal/nrf5/radio/radio.c [hunk 4]
```diff
@@ -646,12 +640,7 @@ u32_t radio_tmr_start(u8_t trx, u32_t ticks_start, u32_t remainder)
 	HAL_EVENT_TIMER_START_PPI_REGISTER_TASK = HAL_EVENT_TIMER_START_TASK;
 	NRF_PPI->CHENSET = HAL_EVENT_TIMER_START_PPI_ENABLE;
 
-	HAL_RADIO_ENABLE_ON_TICK_PPI_REGISTER_EVT =
```
Comment (ioannisg) [inline subsys/bluetooth/controller/hal/nrf5/radio/radio.c]: This is just refactoring.

File: subsys/bluetooth/controller/hal/nrf5/radio/radio.c [hunk 5]
```diff
@@ -665,16 +654,7 @@ u32_t radio_tmr_start(u8_t trx, u32_t ticks_start, u32_t remainder)
 	HAL_SW_SWITCH_TIMER_CLEAR_PPI_REGISTER_TASK =
 		HAL_SW_SWITCH_TIMER_CLEAR_PPI_TASK;
 
-#if defined(CONFIG_SOC_NRF52840)
```
Comment (ioannisg) [inline subsys/bluetooth/controller/hal/nrf5/radio/radio.c]: We use a single PPI for S2 and we don't (by default) assign it to a group, so we don't need different compile-time asignments for 832 and 840.

File: subsys/bluetooth/controller/hal/nrf5/radio/radio.c [hunk 6]
```diff
@@ -690,43 +670,31 @@ u32_t radio_tmr_start(u8_t trx, u32_t ticks_start, u32_t remainder)
...
 	NRF_PPI->CHG[SW_SWITCH_TIMER_TASK_GROUP(1)] =
 		HAL_SW_SWITCH_GROUP_TASK_DISABLE_PPI_1_INCLUDE |
 			HAL_SW_SWITCH_RADIO_ENABLE_PPI_1_INCLUDE;
-#endif /* CONFIG_SOC_NRF52840 */
 #endif /* !CONFIG_BT_CTLR_TIFS_HW */
 
 	return remainder;
 }
 
-static inline void radio_enable_on_timer_tick(u8_t trx)
```
Comment (ioannisg) [inline subsys/bluetooth/controller/hal/nrf5/radio/radio.c]: Refactoring

File: subsys/bluetooth/controller/hal/nrf5/radio/radio_nrf52840.h [hunk 7]
```diff
@@ -66,44 +66,44 @@
 /* TXEN->TXIDLE + TXIDLE->TX (with fast Radio ramp-up mode)
  * in microseconds for LE CODED PHY [S2].
  */
-#define HAL_RADIO_NRF52840_TXEN_TXIDLE_TX_S2_FAST_NS 42300 /* 40.1 + 2.2 */
```
Comment (ioannisg) [inline subsys/bluetooth/controller/hal/nrf5/radio/radio_nrf52840.h]: This patch is added to align with sniffer measurements  on LE Coded.

File: subsys/bluetooth/controller/hal/nrf5/radio/radio_nrf5_ppi.h [hunk 8]
```diff
@@ -26,6 +26,16 @@
 #define HAL_RADIO_ENABLE_ON_TICK_PPI_TASK_RX \
 	((u32_t)&(NRF_RADIO->TASKS_RXEN))
 
+static inline void hal_radio_enable_on_tick_ppi_config_and_enable(u8_t trx)
```
Comment (ioannisg) [inline subsys/bluetooth/controller/hal/nrf5/radio/radio_nrf5_ppi.h]: Refactoring

File: subsys/bluetooth/controller/hal/nrf5/radio/radio_nrf5_ppi.h [hunk 9]
```diff
@@ -305,33 +315,21 @@
 	((u32_t)&(NRF_RADIO->TASKS_RXEN))
 
 #if defined(CONFIG_SOC_NRF52840)
-/* The 2 adjacent TIMER EVENTS_COMPARE event offsets used for implementing
```
Comment (ioannisg) [inline subsys/bluetooth/controller/hal/nrf5/radio/radio_nrf5_ppi.h]: These changes and the ones below are due to the fact that we reserve a single PPI for S2.

**Review by cvinayak — CHANGES_REQUESTED** [review]
Have separate commits for:
1. Introduction of hal_radio_enable_on_tick_ppi_config_and_enable and its use in radio_tmr_start
2. Use of single PPI and CC in sw_switch, and related change in radio_tmr_start
3. Changes to timing constants

Will help in future, to bisect into changes.

File: subsys/bluetooth/controller/hal/nrf5/radio/radio.c (hunk 3 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/hal/nrf5/radio/radio.c]: @ioannisg I remember to have implemented the same (adding and removing the PPI in group), but for reasons unknown I could not get the desired behavior, hence, initialized the group (with 2 sets) once per event, and registered or unregistered in sw_switch.

I see, you have it working with only adding/removing from task group, then I dont see a necessity to explicitly unregistering/reset of the PPI channel.

**Review by cvinayak — APPROVED** [review]

**Pull request merged**